### DISCUSSION TITLE
How perform reset factory Esp32  in code

### DISCUSSION THREAD

**Discussion by TomInIowa** [HomeSpan/HomeSpan#278] [Q&A] [answered]
Teyleten Robot ESP32S ESP32 ESP-WROOM-32 Development Board 2.4GHz Dual-Core WiFi +Bluetooth 2 Function Microcontroller for Arduino (ESP32 30P, 3PCS) from Amazon.

I set pin 23 as the control pin for. homeSpan.
If I set pin 23 low (pinMode(23, OUTPUT); 
digitalWrite(23, LOW);
wait 12 seconds
digitalWrite(23, HIGH);  the ESP32 does NOT react.

If I physical ground pin 23 and then unground it after about 12 seconds, the ESP will restart.
In the monitor window, "*** FACTORY RESET!  Restarting..."

Any idea on why the code does NOT work?

**Comment by HomeSpan** [comment, marked as answer]
The code that reads a pushbutton relies on an initial state of HIGH by setting the mode to INPUT_PULLUP.  Depending on the context in which you changed to OUTPUT, this could have thrown things off.  One other thing I have noticed is that for the ESP32-C3 I use, the above code does not function as expected since setting a pin with mode=OUTPUT does not actually change the reading when accessed as an INPUT.   The solution in this circumstance is to set the pin to INPUT/OUTPUT mode.  This can't be done with the arduino-style pinMode function but it can be done with native esp functions.  However, I have only noticed this on my ESP32-C3 chip, not on an ESP32, which is what you are using above. 

Regardless, there is an easy way to programmatically reset HomeSpan to factory default, though I have not yet documented this in the API.   You can invoke any CLI command using `homeSpan.processSerialCommand(char *);`.  The action will be the same as if you typed the command into the Serial Monitor.  Since typing 'F' into the Serial Monitor resets HomeSpan to factory default, you can simply call `homeSpan.processSerialCommand("F")` to trigger a factory reset.

Hope this helps.

**Reply by TomInIowa** [reply]
It works.   Thank you.    Comes in handy with OTA updates of remote ESP32 board.

**Comment by TomInIowa** [comment]
Thank you!!
I am going to have to look at homeSpan.h to see what other commands are in there.
________________________________
